File: DSA/Bhavya_vyas500121193/Day27_Q2.py
class Node:
    def __init__(self, data): 
        self.data = data
        self.prev = None
        self.next = None

def reverse_doubly_linked_list(head):
    if not head:
        return None

    current = head
    temp = None

    while current:
        temp = current.prev
        current.prev = current.next
        current.next = temp
        current = current.prev  

    if temp:
        head = temp.prev

    return head


def print_list(head):
    while head:
        print(head.data, end=" <-> " if head.next else "\n")
        head = head.next

def build_dll(values):
    head = None
    prev = None
    for val in values:
        node = Node(val)
        if not head:
            head = node
        else:
            prev.next = node
            node.prev = prev
        prev = node
    return head


head = build_dll([75, 122, 59, 196])
print("Before reversal:")
print_list(head)
head = reverse_doubly_linked_list(head)
print("After reversal:")
print_list(head)

File: DSA/Kapil_500119132/Day27/Day27_Q2.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* prev;
    Node* next;
};

Node* reverseDLL(Node* head) {
    Node* temp = nullptr;
    Node* curr = head;
    while (curr) {
        temp = curr->prev;
        curr->prev = curr->next;
        curr->next = temp;
        curr = curr->prev;
    }
    if (temp) head = temp->prev;
    return head;
}
